cmd/process-starter/process_starter.go

```diff
@@ -83,14 +83,16 @@ func cpuListStrToIntSlice(cpuString string) (cpuList []int) {
 	return cpuList
 }
 
-func pollCPUSetCompletion(exclusiveCPUs []int, sharedCPUs []int) {
+func pollCPUSetCompletion()(exclusiveCPUs, sharedCPUs []int) {
```

Comment:
Some comments for the function:

1. Now this function reads the allocated CPUs from environment variables and waits that cpuset is set correctly. Function should exist for one purpose only but maybe the purpose can be changed here. Then the name should be changed accordingly.

2. Does this function work ? The `expCpus` is set only in the beginning of the function when the allocated CPUs are not known (they are read from env variables later) ?

3. This could still fail if e.g only EXCLUSIVE_CPUS is set and cpuset is having only the exclusive cpus but container should get shared cpus also (which it did not get because of timing issue) ?

Reply:
1. Well, I don't think the purpose is changed so much. The correct env variables needs for check cpuset completion.

2. You were right! It doesn't work, but when I tested it looked ok... I don't know how.
Now I've checked it more precisely, and works fine.
But I found an other issue, which is connected to this. The conversation (cpuListStrToIntSlice) cannot handle string like "2-8".

3. I think 10 seconds should be enough to set both EXCLUSIVE_CPUS and SHARED_CPUS.

Reply:
For the 3rd one I'm thinking a case where EXCLUSIVE_CPUS is set but SHARED_CPUS is not set yet. And at that moment cgroup cpuset is set to contain only exclusive cpus.  In that case we are returning from the function (not waiting the whole 10 seconds). CPU affinity setting should fail then as cpuset does not contain shared cpus ?

Reply:
I definitely don't think we should add additional sleeps
I'm also not sure if we even want to handle that very specific edge cases until that happens

proposal: always take the union of EXCLUSIVE_CPUS and SHARED_CPUS, if one of them is missing then it would be counted as "empty set". Then compare it against the container's cpuset
I think it is extremely unlikely that the union of the tow environment variables is not fully provisioned and is **exactly matching** the cpuset of a container which is also not yet fully provisioned

Reply:
It was quite easy for me to get incomplete cpuset hence the polling was introduced. I did not see the environment variable problem but if it is easy to reproduce, the probability to see incomplete cpuset **and** missing environment variable shouldn't be so small. 

I agree that we shouldn't implement more sleeps but could we try to find out what is expected and then act based on that. Webhook could set an environment variable indicating the pool allocations (pool types and maybe cpu amounts). Based on this variable the process starter knows what to expect.

Reply:
that sounds a great idea! a third env variable would solve the issue

Reply:
ok, so this actually does not work
we are sending back environment variables in the Allocate() response struct, but that is happening per device if I'm not mistaken. so without maintaining internal state the DP should not be aware what additional devices were, or will be allocated to the same container
so in any given  Allocate response we always only send back max 1 env variable
BTW probably this is the reason why they are asynchronously provisioned :) I guess  Kublete pushes them as soon as the Allocate responses arrive

Point2: cpusetter provisions the containers cpuset  in one go. So we either have the default provisioned, or the final.
It can never be the case that a partially provisioned env variable state matches with a partially provisioned cpuset, because we never provisione the cpuset partially.
It can however theoretically match with the default, Kubelet provisioned cpuset, but that is really an insignificant chance IMO; considering our pools are distinct (not sure if we validate this though)

Point3: if we want to be sure, we could however provision a third environment variable from the Webhook!
A: Webhook already has the Pod spec, and can patch changes into it
B: webhook does not know the value of the future env variables in creation time, however it can tell how many will be created (0,1, or 2)
Webhook could create a third env variable in the container conveying this number. Based on this number process starter could determine how many variables it should wait for before checking the cpuset

Reply:
If it is guaranteed that at least one pool allocation is completed and related environment variable is set when container starts, we should be fine with current code. If not, I would implement the webhook part also. It's quite small task anyway and it adds robustness.

Reply:
I implemented the webhook part. Please check it, if it is ok for you.

Reply:
I think it is good now. If there is nothing else coming, we can merge this.

Reply:
Thanks! I don't think any other thing needed. Maybe Levo has some comments...